man/rasdaemon.1.in
.\"****************************************************************************
.\" $Id$
.\"****************************************************************************
.\" SPDX-License-Identifier: GPL-2.0-or-later
.\"
.\"Copyright (c) 2013 Mauro Carvalho Chehab <mchehab+huawei@kernel.org>
.\"
.\" This is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This is distributed in the hope that it will be useful, but WITHOUT
.\" ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
.\" FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA.
.\"****************************************************************************

.TH RASDAEMON 8 "@META_DATE@" "@META_ALIAS@" "RAS memory controller admin utility"

.SH NAME
rasdaemon \- RAS daemon to log the RAS events.

.SH SYNOPSIS
.B rasdaemon
[\fIOPTION\fR]...

.SH DESCRIPTION

The \fBrasdaemon\fR program is a daemon which monitors the platform
Reliablity, Availability and Serviceability (RAS) reports from the
Linux kernel trace events. These trace events are logged in
/sys/kernel/debug/tracing, reporting them via syslog/journald. If
syslog-ng is installed, the events will logged at @localstatedir@/log/rasdaemon.

.SH OPTIONS
.TP
.BI "--usage"
Display a brief usage message and exit.
.TP
.BI "--help"
Display a help message and exit.
.TP
.BI "--disable"
Disable RAS tracing events and exit.
.TP
.BI "--enable"
Enable RAS tracing events and exit.
.TP
.BI "--foreground"
Executes in foreground, printing the events at console. Useful for testing it,
and to be used by systemd or Unix System V respan.
If not specified, the program runs in daemon mode.
.TP
.BI "--record"
Record RAS events via Sqlite3. The Sqlite3 database has the benefit of
keeping a persistent record of the RAS events. This feature is used with
the ras-mc-ctl utility. Note that rasdaemon may be compiled without this
feature.
.TP
.BI "--version"
Print the program version and exit.

.SH CONFIG FILE

The \fBrasdaemon\fR program supports a config file to set rasdaemon systemd service
environment variables. By default the config file is read from /etc/sysconfig/rasdaemon.

The general format is environmentname=value.

.SH LOG LEVEL

Each log entry has a level prefix that describes the severity of the log to
help users determine which logs are more valuable.
Currently, three levels are used:.TP

.B "ALERT"
The uncorrected hardware error has been fixed, but cause side effects.
.TP
.B "CRIT"
The uncorrected hardware error has been detected.
.TP
.B "ERROR"
The corrected hardware error has been detected.

.SH SEE ALSO
\fBras-mc-ctl\fR(8)



NEWS
RAS DAEMON
==========

In Kernel 3.5 we've started to address the long-discussed need of having a
better way to handle platform Reliability, Availability and Serviceability
(RAS).

Basically, a tracepoint event that handles memory errors called ras:mc_event
was added there, together with HERM/EDAC version 3.0 patches.

In Kernel 3.8, a new event was added, to handle PCIe AER events (ras:aer_event)
[1].

On kernel 3.9, a new driver was added to report hardware memory errors that
comes from the BIOS via ras:mc_event (the new ghes_edac driver).

It is still on my TODO list to add a RAS trace event for non-memory related
errors that come via the MCA machine check handler (mcelog).

While progress made was made at Kernel infrastructure, the needed userspace
tools were still lacking. So, I decided to start materializing the userspace
counterpart for what it was informally named as rasdaemon on some discussions.

The rasdaemon tool is available at:
	http://git.infradead.org/users/mchehab/rasdaemon.git

The current version is on very early stages, and it has a copy on it of
the library that Steven Rostedt's is writing for trace-cmd tool. The plan
is to use the trace-cmd library, when it starts to packaged as a separate
library. I'd like to thanks Steven for the help he gave me to write this
initial version.

The current version of the tool enables the ras:mc_event log, and reads
it via the raw trace debugfs node:
	/sys/kernel/debug/tracing/per_cpu/cpu*/trace_pipe_raw

It also has a code that allows recording the errors via an sqlite3 database.

The long term plan is to provide a tool that will catch and handle all
ras:* error events that comes from the Kernel tracing infrastructure,
logging them and providing tools to report it, being able to detect burst
errors (like the ones caused by a solar storm at memories) or sparsed errors,
in a way that would provide a glue to the users about the root cause of the
error.

Of course, there are much to do there.

It is a natural evolution of the tool to add support there for the
ras:aer_event traces that can come from PCIe AER.

While it currently works with current Kernels since kernel 3.5, there are a
number of interesting changes at tracing that are planned to be merged for
Kernel 3.10:
	- poll() support for per_cpu trace_pipe_raw;
	- a timestamp that could more easily associated with machine's
	  uptime;
	- support for a separate ringbuffer for RAS.

So, it is planned the minimal requirement for the final version (v1.0)
would be kernel 3.10.

This is currently on very early staging. Help is needed ;)

So, please send us suggestions, patches etc to the EDAC mailing list:
	linux-edac@vger.kernel.org

Thanks,
Mauro Carvalho Chehab
2013-03-14

-

[1] Currently, ras:mc_event is at include/ras/. It is on my todo list
to move it to be together with ras:aer_event, at include/trace/events/ras.h.


rasdaemon.c
// SPDX-License-Identifier: GPL-2.0-or-later

/*
 * Copyright (C) 2013 Mauro Carvalho Chehab <mchehab+huawei@kernel.org>
 */

#include <argp.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/syslog.h>
#include <unistd.h>

#include "ras-events.h"
#include "ras-logger.h"
#include "ras-record.h"
#include "ras-pcie-edpc.h"
#include "types.h"

/*
 * Arguments(argp) handling logic and main
 */

#define TOOL_NAME "rasdaemon"
#define TOOL_DESCRIPTION "RAS daemon to log the RAS events."
#define ARGS_DOC "<options>"
#define DISABLE "DISABLE"

const char *argp_program_version = TOOL_NAME " " VERSION;
const char *argp_program_bug_address = "Mauro Carvalho Chehab <mchehab@kernel.org>";

struct arguments {
	int record_events;
	int enable_ras;
	int enable_ipmitool;
	int foreground;
	int offline;
};

enum OFFLINE_ARG_KEYS {
	SMCA = 0x100,
	MODEL,
	FAMILY,
	BANK_NUM,
	IPID_REG,
	STATUS_REG,
	SYNDROME_REG
};

struct ras_mc_offline_event event;

static error_t parse_opt(int k, char *arg, struct argp_state *state)
{
	struct arguments *args = state->input;

	switch (k) {
	case 'e':
		args->enable_ras++;
		break;
	case 'd':
		args->enable_ras--;
		break;
#ifdef HAVE_SQLITE3
	case 'r':
		args->record_events++;
		break;
#endif
#ifdef HAVE_OPENBMC_UNIFIED_SEL
	case 'i':
		args->enable_ipmitool++;
		break;
#endif
	case 'f':
		args->foreground++;
		break;
#ifdef HAVE_MCE
	case 'p':
		if (state->argc < 4)
			argp_state_help(state, stdout, ARGP_HELP_LONG | ARGP_HELP_EXIT_ERR);
		args->offline++;
		break;
#endif
	default:
		return ARGP_ERR_UNKNOWN;
	}
	return 0;
}

#ifdef HAVE_MCE
static error_t parse_opt_offline(int key, char *arg,
				 struct argp_state *state)
{
	switch (key) {
	case SMCA:
		event.smca = true;
		break;
	case MODEL:
		event.model = strtoul(state->argv[state->next], NULL, 0);
		break;
	case FAMILY:
		event.family = strtoul(state->argv[state->next], NULL, 0);
		break;
	case BANK_NUM:
		event.bank = atoi(state->argv[state->next]);
		break;
	case IPID_REG:
		event.ipid = strtoull(state->argv[state->next], NULL, 0);
		break;
	case STATUS_REG:
		event.status = strtoull(state->argv[state->next], NULL, 0);
		break;
	case SYNDROME_REG:
		event.synd = strtoull(state->argv[state->next], NULL, 0);
		break;
	default:
		return ARGP_ERR_UNKNOWN;
	}
	return 0;
}
#endif

long user_hz;

int main(int argc, char *argv[])
{
	struct arguments args;
	int idx = -1;

	choices_disable = getenv(DISABLE);

#ifdef HAVE_MCE
	const struct argp_option offline_options[] = {
		{"smca", SMCA, 0, 0, "AMD SMCA Error Decoding"},
		{"model", MODEL, 0, 0, "CPU Model"},
		{"family", FAMILY, 0, 0, "CPU Family"},
		{"bank", BANK_NUM, 0, 0, "Bank Number"},
		{"ipid", IPID_REG, 0, 0, "IPID Register (for SMCA systems only)"},
		{"status", STATUS_REG, 0, 0, "Status Register"},
		{"synd", SYNDROME_REG, 0, 0, "Syndrome Register"},
		{0, 0, 0, 0, 0, 0},
	};

	struct argp offline_argp = {
		.options = offline_options,
		.parser = parse_opt_offline,
		.doc = TOOL_DESCRIPTION,
		.args_doc = ARGS_DOC,
	};

	struct argp_child offline_parser[] = {
		{&offline_argp, 0, "Post-Processing Options:", 0},
		{0, 0, 0, 0},
	};
#endif

	const struct argp_option options[] = {
		{"enable",  'e', 0, 0, "enable RAS events and exit", 0},
		{"disable", 'd', 0, 0, "disable RAS events and exit", 0},
#ifdef HAVE_SQLITE3
		{"record",  'r', 0, 0, "record events via sqlite3", 0},
#endif
		{"foreground", 'f', 0, 0, "run foreground, not daemonize"},
#ifdef HAVE_OPENBMC_UNIFIED_SEL
		{"ipmitool", 'i', 0, 0, "enable ipmitool logging", 0},
#endif
#ifdef HAVE_MCE
		{"post-processing", 'p', 0, 0,
		"Post-processing MCE's with raw register values"},
#endif

		{ 0, 0, 0, 0, 0, 0 }
	};
	const struct argp argp = {
		.options = options,
		.parser = parse_opt,
		.doc = TOOL_DESCRIPTION,
		.args_doc = ARGS_DOC,
#ifdef HAVE_MCE
		.children = offline_parser,
#endif
	};
	memset(&args, 0, sizeof(args));

	user_hz = sysconf(_SC_CLK_TCK);

	argp_parse(&argp, argc, argv, 0,  &idx, &args);

	if (idx < 0) {
		argp_help(&argp, stderr, ARGP_HELP_STD_HELP, TOOL_NAME);
		return -1;
	}

	if (args.enable_ras) {
		int enable;

		enable = (args.enable_ras > 0) ? 1 : 0;
		toggle_ras_mc_event(enable);

		return 0;
	}

#ifdef HAVE_MCE
	if (args.offline) {
		ras_offline_mce_event(&event);
		return 0;
	}
#endif

	openlog(TOOL_NAME, 0, LOG_DAEMON);
	if (!args.foreground)
		if (daemon(0, 0))
			exit(EXIT_FAILURE);

	if (getenv(PCIE_EDPC_ENABLE) && atoi(getenv(PCIE_EDPC_ENABLE)))
		config_pcie_edpc();
	else
		log(TERM, LOG_INFO, "PCIE EDPC config is not enabled\n");

	handle_ras_events(args.record_events, args.enable_ipmitool);

	return 0;
}
